src/v/iceberg/tests/metadata_query_test.cc

```diff
@@ -0,0 +1,419 @@
+    };
+
+    auto [snapshots, files, manifests] = execute_for_all_results(
+      tx.table(), std::nullopt, matcher);
+
+    // it should only contain snapshots 3,4,5
+    ASSERT_EQ(snapshots.size(), 3);
+    EXPECT_THAT(
+      snapshots,
+      Each(Field(&snapshot::sequence_number, Ge(sequence_number(3)))));
+
+    auto all_manifest_files = collect_all_manifest_files(tx.table()).get();
+
+    ASSERT_EQ(files.size(), 3);
+    ASSERT_EQ(files[0], all_manifest_files[2]);
+    ASSERT_EQ(files[1], all_manifest_files[3]);
+    ASSERT_EQ(files[2], all_manifest_files[4]);
+
+    auto all_manifests = collect_all_manifests(tx.table()).get();
+    ASSERT_EQ(manifests.size(), 3);
+    ASSERT_EQ(manifests[0], all_manifests[2]);
+    ASSERT_EQ(manifests[1], all_manifests[3]);
+    ASSERT_EQ(manifests[2], all_manifests[4]);
+}
+
+TEST_F(MetadataQueryTest, TestQueryManifests) {
+    auto tx = generate_metadata().get();
+
+    auto matcher = [](const manifest& m) { return m.entries.size() == 3; };
+    metadata_query<result_type::snapshot> snap_query{
+      .manifest_matcher = matcher};
+    metadata_query<result_type::manifest_file> mf_query{
+      .manifest_matcher = matcher};
+    metadata_query<result_type::manifest> m_query{.manifest_matcher = matcher};
+
+    auto [snapshots, files, manifests] = execute_for_all_results(
+      tx.table(), std::nullopt, std::nullopt, matcher);
+    // it should only contain 3 snapshots as they contain manifest we are
+    // looking for
+    ASSERT_EQ(snapshots.size(), 3);
+    ASSERT_EQ(snapshots[0], (*tx.table().snapshots)[2]);
+    ASSERT_EQ(snapshots[1], (*tx.table().snapshots)[3]);
+    ASSERT_EQ(snapshots[2], (*tx.table().snapshots)[4]);
+
+    auto all_manifest_files = collect_all_manifest_files(tx.table()).get();
+
+    // only one manifest has 3 entries, we deduplicate the result so here
+    // there should be only one element
+    ASSERT_EQ(files.size(), 1);
+    ASSERT_EQ(files[0], all_manifest_files[2]);
+
+    auto all_manifests = collect_all_manifests(tx.table()).get();
+    ASSERT_EQ(manifests.size(), 1);
+    ASSERT_EQ(manifests[0], all_manifests[2]);
+}
+
+template<typename TupleT, std::size_t... Is>
+AssertionResult do_check_query_result_is_empty(
+  const query_executor& executor,
+  const TupleT& tp,
+  std::index_sequence<Is...>) {
+    auto query = [&executor](const auto& query) {
+        auto r = executor.execute_query(query).get();
+
+        return r.has_value() && r.value().empty();
+    };
+    bool success = (query(std::get<Is>(tp)) && ...);
+    if (!success) {
+        return AssertionFailure() << "Queries should return empty results";
+    }
+    return AssertionSuccess();
+}
+
+template<typename TupleT, std::size_t tuple_size = std::tuple_size_v<TupleT>>
+AssertionResult
+check_query_result_is_empty(const query_executor& executor, const TupleT& tp) {
+    return do_check_query_result_is_empty(
+      executor, tp, std::make_index_sequence<std::tuple_size_v<TupleT>>{});
+}
```

Comment:
Feels like a lot of machinery for what's being tested, but thanks for keeping the test readable :)